File: src/year2018/day16/day16.rs
//! [Day 16: Chronal Classification](https://adventofcode.com/2018/day/16)

use itertools::Itertools;
use rustc_hash::{FxHashMap, FxHashSet};

const OPCODES: [&str; 16] = [
    "addi", "addr", "bani", "banr", "bori", "borr", "eqir", "eqri", "eqrr", "gtir", "gtri", "gtrr",
    "muli", "mulr", "seti", "setr",
];

fn emulate(opcode: &str, a: u32, b: u32, c: u32, regs: &[u32]) -> Vec<u32> {
    let mut output = regs.to_vec();

    output[c as usize] = match opcode {
        "addr" => regs[a as usize] + regs[b as usize],
        "addi" => regs[a as usize] + b,
        "mulr" => regs[a as usize] * regs[b as usize],
        "muli" => regs[a as usize] * b,
        "banr" => regs[a as usize] & regs[b as usize],
        "bani" => regs[a as usize] & b,
        "borr" => regs[a as usize] | regs[b as usize],
        "bori" => regs[a as usize] | b,
        "setr" => regs[a as usize],
        "seti" => a,
        "gtir" => u32::from(a > regs[b as usize]),
        "gtri" => u32::from(regs[a as usize] > b),
        "gtrr" => u32::from(regs[a as usize] > regs[b as usize]),
        "eqir" => u32::from(a == regs[b as usize]),
        "eqri" => u32::from(regs[a as usize] == b),
        "eqrr" => u32::from(regs[a as usize] == regs[b as usize]),
        _ => panic!("bad opcode {opcode}"),
    };

    output
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let (samples, program) = data.split_once("\n\n\n\n").unwrap();

    // the mapping between opcode numebrs and labels
    let mut mapping: FxHashMap<u32, FxHashSet<&str>> = FxHashMap::default();

    // parse the CPU monitoring
    let parse_line = |line: &str, prefix: &str| -> Vec<u32> {
        line.strip_prefix(prefix)
            .unwrap()
            .strip_suffix(']')
            .unwrap()
            .split(',')
            .map(str::trim)
            .map(|x| x.parse().unwrap())
            .collect()
    };

    // parse the samples
    let mut result1 = 0;

    for sample in samples.split("\n\n") {
        let sample: Vec<_> = sample.lines().collect();

        let before = parse_line(sample[0], "Before: [");
        let after = parse_line(sample[2], "After:  [");

        let instruction: Vec<_> = sample[1]
            .split_ascii_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();

        let (o, a, b, c) = instruction.iter().copied().collect_tuple().unwrap();

        // find opcodes that transform 'before' to 'after'
        let matches = OPCODES
            .iter()
            .filter(|opcode| emulate(opcode, a, b, c, &before) == after)
            .map(|opcode| {
                mapping.entry(o).or_default().insert(opcode);
            })
            .count();

        // compute part 1
        if matches >= 3 {
            result1 += 1;
        }
    }

    // resolve the mapping
    let mut opcodes = FxHashMap::default();

    while !mapping.is_empty() {
        // find the opcode with one label exactly (or panic, but puzzle inputs are built to prevent this)
        let (opcode, labels) = mapping
            .iter()
            .find(|(_, labels)| labels.len() == 1)
            .unwrap();

        // dereference the found opcode/label
        let opcode = *opcode;
        let label = *labels.iter().next().unwrap();

        // remove it from the mapping
        mapping.remove(&opcode);
        for i in mapping.values_mut() {
            i.remove(label);
        }

        // and insert it into the opcode resolver
        opcodes.insert(opcode, label);
    }

    // run the test program
    let mut regs = vec![0; 4];

    for instruction in program.lines() {
        let instruction: Vec<_> = instruction
            .split_ascii_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();

        let (o, a, b, c) = instruction.iter().copied().collect_tuple().unwrap();

        regs = emulate(opcodes.get(&o).unwrap(), a, b, c, &regs);
    }

    // part 2 is the content of register 0
    let result2 = regs[0];

    (result1, result2)
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

File: src/year2018/day16/day16.py
#!/usr/bin/env python3
# [Day 16: Chronal Classification](https://adventofcode.com/2018/day/16)

import atexit
import sys
import time
import typing as t
from collections import defaultdict
from copy import deepcopy
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
self_tests = "-T" in sys.argv
if self_tests:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


OPCODES = (
    "addi",
    "addr",
    "bani",
    "banr",
    "bori",
    "borr",
    "eqir",
    "eqri",
    "eqrr",
    "gtir",
    "gtri",
    "gtrr",
    "muli",
    "mulr",
    "seti",
    "setr",
)


def emulate(opcode, a, b, c, regs: t.List) -> t.List:
    regs = deepcopy(regs)
    match opcode:
        # Addition
        case "addr":
            regs[c] = regs[a] + regs[b]
        case "addi":
            regs[c] = regs[a] + b

        # Multiplication
        case "mulr":
            regs[c] = regs[a] * regs[b]
        case "muli":
            regs[c] = regs[a] * b

        # Bitwise AND
        case "banr":
            regs[c] = regs[a] & regs[b]
        case "bani":
            regs[c] = regs[a] & b

        # Bitwise OR
        case "borr":
            regs[c] = regs[a] | regs[b]
        case "bori":
            regs[c] = regs[a] | b

        # Assignment
        case "setr":
            regs[c] = regs[a]
        case "seti":
            regs[c] = a

        # Greater-than testing
        case "gtir":
            regs[c] = 1 if a > regs[b] else 0
        case "gtri":
            regs[c] = 1 if regs[a] > b else 0
        case "gtrr":
            regs[c] = 1 if regs[a] > regs[b] else 0

        # Equality testing
        case "eqir":
            regs[c] = 1 if a == regs[b] else 0
        case "eqri":
            regs[c] = 1 if regs[a] == b else 0
        case "eqrr":
            regs[c] = 1 if regs[a] == regs[b] else 0

        case _:
            raise ValueError(f"bad opcode {opcode}")

    return regs


samples, program = data.split("\n\n\n\n", maxsplit=1)


mapping = defaultdict(set)

result = 0
for sample in samples.split("\n\n"):
    before, instruction, after = sample.splitlines()

    before = list(map(int, before.removeprefix("Before: [").removesuffix("]").split(",")))
    opcode, a, b, c = map(int, instruction.split())
    after = list(map(int, after.removeprefix("After:  [").removesuffix("]").split(",")))

    n = 0
    for o in OPCODES:
        if after == emulate(o, a, b, c, before):
            n += 1
            mapping[opcode].add(o)
    if n >= 3:
        result += 1

# part 1
print(result)

# resolve the mapping
opcodes = dict()
while mapping:
    for k, v in mapping.items():
        if len(v) == 1:
            name = v.pop()
            mapping.pop(k)
            opcodes[k] = name
            for _, v in mapping.items():
                v.discard(name)
            break
    else:
        raise ValueError

# run the test program
regs = [0, 0, 0, 0]
for instruction in program.splitlines():
    opcode, a, b, c = map(int, instruction.split())
    regs = emulate(opcodes[opcode], a, b, c, regs)

# part 2
print(regs[0])
